### PR TITLE
Added ConvexPolygon::expand and Rectangle::expand from vector

### PR DESCRIPTION
**Pull request by jonathanlew**
### Description
Looking at where Rectangle::expand is being used, I observed that we can can expand in particular
directions, for example: expanding a defense area should be expanded towards the front and sides and not expanded into the field boundaries. ConvexPolygon::expand will give us more flexibility and precision and we can integrate this into future work on #1331 once #812 is done

### Testing Done
unit tests

### Resolved Issues
Creates the utils for #1331

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/new_geom/convex_polygon.cpp
Comment (MathewMacDougall) [inline src/software/new_geom/convex_polygon.cpp]: Because this is geometry, I need to ask if `==` is numerically fragile here? May want to do a comparison using `GeomConstants::EPSILON`

File: src/software/new_geom/convex_polygon.h
Comment (MathewMacDougall) [inline src/software/new_geom/convex_polygon.h]: super nit on my preferred wording
```suggestion
     * Returns the convex polygon expanded in the direction of v
```

File: src/software/new_geom/convex_polygon.h
Comment (MathewMacDougall) [inline src/software/new_geom/convex_polygon.h]: nit to be consistent with the rest of the description
```suggestion
     * @return a convex polygon expanded in the direction of v
```

File: src/software/new_geom/convex_polygon.h
Comment (MathewMacDougall) [inline src/software/new_geom/convex_polygon.h]: Since this is implemented in `ConvexPolygon`, other "concrete shapes" we have classes for like `Rectangle` also have this function.

Currently (at least IMO) rectangles are not defined with the concept of rotation, ie. their edges always lie along the x or y axis.

The expansion function would allow rectangles to be expanded diagonally to form convex polygons, but what may no longer be considered a rectangle because it ends up rotated / skewed etc.

Edit: realized this is not a virtual function so is `ConvexPolygon` only. This raises a new question: Seems like we may move to replace `Rectangle::expand` with this function because it's more useful. Do we even have a good use-case for a specific `Rectangle` class anymore? We definitely might just pointing this out.

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]

File: src/software/new_geom/convex_polygon.h
Comment (jonathanlew) [reply src/software/new_geom/convex_polygon.h]: Rectangles are useful because they are an abstraction around construction (give 2 points instead of 4 and guaranteed shape) and for accessing the corners (e.g. negxnegy...). It raises a good point: we don't need `Triangle` anymore now that we have `centroid`, but I think it's still nice to have for readability

File: src/software/new_geom/convex_polygon.cpp
Comment (jonathanlew) [reply src/software/new_geom/convex_polygon.cpp]: OK I changed it to use orientation (which side of the line perpendicular to v along c). The only fp error issue is if the point is close to the line, which shouldn't be a problem because it will decide one way or the other and "double down" on that decision (either by shifting the point even further away from the centroid, or by leaving the point as is and pulling the centroid away from the point
https://stackoverflow.com/questions/1560492/how-to-tell-whether-a-point-is-to-the-right-or-left-side-of-a-line

File: src/software/new_geom/convex_polygon.h
Comment (MathewMacDougall) [reply src/software/new_geom/convex_polygon.h]: I don't mind having `Triangle` because it's a very "unconstrained" version of a `ConvexPolygon` (ie. any 3 points make a triangle basically). I think it helps readability at the very least.

On the other hand, `Rectangle` is more constrained in that the instant you skew it, it's technically no longer a rectangle. I agree the interface with the corners for creation and stuff is useful.

Just off the top of my head though, it seems like the lines between `Rectangles` and `ConvexPolygons` get blurred. We represent some parts of the field (the entire boundary, etc.) with traditional `Rectangles`, but now would want to represent other parts like the defense areas with `ConvexPolygons` because they have a more convenient `expand` interface for how they are used (ie. in avoid areas).

just feels weird to have shapes and their usages intermingled like that, but "feels" is not a very strong argument for anything.

Would definitely like opinions @jonathanlew @garethellis0 @akhilveeraghanta @jsn0716 @scveloso 

Possible solutions:
* Remove `Rectangle` and just use `ConvexPolygon`
* Make `expand` virtual and have `Rectangle` override it. Same interface but clamps vectors to one of the 4 cardinal directions
* Do not make `expand` virtual but implement a separate version for rectangles that takes an enum for the direction and the magnitude. eg. `expand(NEGATIVE_X, 1.2)`

Regardless, I do like the changes in this PR.

Comment (jonathanlew) [reply src/software/new_geom/convex_polygon.h]: Breaking the invariants for `Rectangle` is a really good point. Here's an idea: for a given `expand` vector `v`, break it up into `x` and `y` components and `expand` one after the other. For example, if `v = (-1,2)` then first `expand` with `v = (-1,0)` then `expand` with `v = (0,2)`. That should maintain the invariants of `Rectangle`, and it might make the code look nicer.

**Review by garethellis0 — CHANGES_REQUESTED** [review]

File: src/software/new_geom/convex_polygon.h
Comment (garethellis0) [reply src/software/new_geom/convex_polygon.h]: I don't really see the issue with this function as is? As long as it has no side effects, it's returning a `ConvexPolygon`, so we never create an invalid `Rectangle`?

**Review by akhilveeraghanta — COMMENTED** [review]
LGTM, pending other comments

File: src/software/new_geom/convex_polygon.h
Comment (jonathanlew) [reply src/software/new_geom/convex_polygon.h]: That's true, there's nothing wrong with the function as is, but `expand` is currently used primarily by `Rectangle` so I'll write `Rectangle expand(void)` with the idea above

File: src/software/new_geom/convex_polygon.cpp
Comment (jonathanlew) [reply src/software/new_geom/convex_polygon.cpp]: added

File: src/software/new_geom/convex_polygon.h
Comment (MathewMacDougall) [reply src/software/new_geom/convex_polygon.h]: Yea @garethellis0 this was not an issue with the current function as-is, more of a note that we have different use-cases and different `expand` functions for the different geometry types.

**Review by MathewMacDougall — APPROVED** [review]
LGTM, I like the new rectangle changes. Just checking, we'll remove the old `Rectangle::expand` in a separate PR / issue?

**Comment by jonathanlew** [comment]
> LGTM, I like the new rectangle changes. Just checking, we'll remove the old `Rectangle::expand` in a separate PR / issue?

Yes, that'll be in #1331 or #1332

**Review by garethellis0 — DISMISSED** [review]

File: src/software/new_geom/convex_polygon.cpp
Comment (EvanMorcom) [inline src/software/new_geom/convex_polygon.cpp:238]: I'm thinking LEFT and RIGHT might be confusing without context.

Is there anything else we could name these to make them more descriptive?

From what I see they are 'left' and 'right' of the direction of the Vector, but not necessarily the field?
@garethellis0 @MathewMacDougall

File: src/software/new_geom/convex_polygon.cpp
Comment (EvanMorcom) [inline src/software/new_geom/convex_polygon.cpp]: I'd like to see a short comment there explaining what it happening in the if block below.

The geom stuff is a little difficult to read due to the variable names

File: src/software/new_geom/convex_polygon.h
Comment (EvanMorcom) [inline src/software/new_geom/convex_polygon.h]: Not super familiar with the polygon class, but would it be more accurate to say 'split along the centroid of the shape' than 'in half'?

**Review by EvanMorcom — CHANGES_REQUESTED** [review]
LGTM 🚋 :fire:

Couple comments

File: src/software/new_geom/convex_polygon.cpp
Comment (garethellis0) [reply src/software/new_geom/convex_polygon.cpp:238]: I can see where there might have been some confusion here. Perhaps it's worth adding a comment clarifying that "left" and "right" are with respect to the vector pointing "straight up".

File: src/software/new_geom/convex_polygon.h
Comment (jonathanlew) [reply src/software/new_geom/convex_polygon.h]: Yeah that's right. Looking at how I implemented `expand` for `Rectangles` (first expand x component, then y component), I think I should do the same for `ConvexPolygon` because the implementations would be consistent and we won't have to reason about shapes being cut along weird axes. @EvanMorcom @garethellis0 @MathewMacDougall What do you think?

File: src/software/new_geom/convex_polygon.cpp
Comment (MathewMacDougall) [inline src/software/new_geom/convex_polygon.cpp]: This looks like the cross-product. We have a `Vector::cross` function that would make this more readable, and a more descriptive variable name would help too.

File: src/software/new_geom/convex_polygon.h
Comment (MathewMacDougall) [reply src/software/new_geom/convex_polygon.h]: The rectangle is still split before it's expanded, it's just easier to reason about since it's a more well-defined shape.

@jonathanlew are you just suggesting applying the "expansion" to each point in the same way as rectangle (ie. by x/y component individually) without any change to _what_ points are expanded?

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]

File: src/software/new_geom/convex_polygon.h
Comment (jonathanlew) [reply src/software/new_geom/convex_polygon.h]: @MathewMacDougall There is a difference between the two types of expansion. Really wish I could whiteboard this but here's a beautiful paint drawing to explain
image

Comment (EvanMorcom) [reply src/software/new_geom/convex_polygon.h]: I feel like the bottom case would be the more expected behaviour IMO

Comment (jonathanlew) [reply src/software/new_geom/convex_polygon.h]: Then that's a problem for Rectangle::expand. I'll change that to only do vertical and horizontal

File: src/software/new_geom/convex_polygon.cpp
Comment (jonathanlew) [reply src/software/new_geom/convex_polygon.cpp]: I've updated the variable names

File: src/software/new_geom/convex_polygon.cpp
Comment (jonathanlew) [reply src/software/new_geom/convex_polygon.cpp:238]: updated the comments

File: src/software/new_geom/convex_polygon.cpp
Comment (jonathanlew) [reply src/software/new_geom/convex_polygon.cpp]: refactored with cross

File: src/software/new_geom/convex_polygon.h
Comment (MathewMacDougall) [reply src/software/new_geom/convex_polygon.h]: Oh nice drawing @jonathanlew . I didn't realize in the ConvexPolygon case we'd take the points on the "side" of the centroid for each component.

**Review by MathewMacDougall — DISMISSED** [review]
LGTM as long as everyone else agrees these interfaces / behaviour for `expand` makes sense

**Review by garethellis0 — CHANGES_REQUESTED** [review]

File: src/software/new_geom/rectangle.h
Comment (jonathanlew) [reply src/software/new_geom/rectangle.h]: I don't want to deal with shrinking anything so the interface cannot allow things to shrink. I rather not resort to throwing std::invalid_argument if possible

Comment (jonathanlew) [reply src/software/new_geom/rectangle.h]: What if I change `expand(double amount)` to `legacyAdditiveSizeChange` and make it deprecated?

File: src/software/new_geom/convex_polygon.h
Comment (EvanMorcom) [reply src/software/new_geom/convex_polygon.h]: Wait, so which method is used for ConvexPolygon and Rectangle?

File: src/software/new_geom/rectangle.h
Comment (EvanMorcom) [reply src/software/new_geom/rectangle.h]: I'm not sure if I agree that a rectangle with a negative expansion should shrink. I think a .shrink() more explicitly defines that behaviour.

There are cases when the sign of the expansion (and there for the direction of expansion of the rectangle) matters.

@garethellis0 @MathewMacDougall

**Review by EvanMorcom — COMMENTED** [review]
Still LGTM 💵 

Just waiting on comment resolution before approval

File: src/software/new_geom/rectangle.h
Comment (jonathanlew) [reply src/software/new_geom/rectangle.h]: I think the problem is that `expand(double amount)` had shrinking behaviour and having functions with the same name in the same class have shrinking and non-shrinking behaviours is confusing. I hope renaming things to  `legacyAdditiveSizeChange`, `expandXMinOrXMax`, and `expandYMinOrYMax` makes things clearer

Comment (MathewMacDougall) [reply src/software/new_geom/rectangle.h]: IMO `expandXMinOrXMax` is still a bit of a confusing name. Upon reading this again, I agree that I would definitely not assume a negative number changes which points are expanded in which direction. Because we can't have `unsigned doubles`, it might be impossible to completely avoid someone accidentally shrinking something.

Comment (jonathanlew) [reply src/software/new_geom/rectangle.h]: After some offline discussion, we decided that `Rectangle::expand` will expand one of the corners by `v`.

**Review by MathewMacDougall — APPROVED** [review]
LGTM :+1:

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada: - nice job

**Review by EvanMorcom — APPROVED** [review]
LGTM :bread: 🥯

**Pull request merged**
